source/SH3/graphics/texture.cpp

```diff
@@ -9,52 +9,380 @@
  *  @author Jesse Buhagiar
  */
 #include <SH3/graphics/texture.hpp>
+#include <SH3/system/assert.hpp>
 #include <SH3/system/log.hpp>
 #include <SH3/arc/vfile.hpp>
 
+#include <algorithm>
 #include <cassert>
-
+#include <fstream>
+#include <functional>
+#include <limits>
 
 using namespace sh3_graphics;
 
+#pragma pack(push, 1)
+struct rgba
+{
+    std::uint8_t r, g, b, a;
+};
+
+struct rgba16
+{
+    std::uint8_t r : 5;
+    std::uint8_t g : 5;
+    std::uint8_t b : 5;
+    std::uint8_t a : 1;
+};
+
+struct rgb
+{
+    std::uint8_t r, g, b;
+};
+
+struct bgra
+{
+    std::uint8_t b, g, r, a;
+};
+
+struct bgr
+{
+    std::uint8_t b, g, r;
+};
+
+struct tga_header
+{
+    static constexpr std::uint8_t TYPE_RGB24 = 2;
+    static constexpr std::uint8_t FLAGS_FLIP = 0x20;
+
+    std::uint8_t id_size = 0;               /**< Size of the ID field that follows this header (usually 0) */
+    std::uint8_t colormap = 0;              /**< Is this image paletted (THIS MUST ALWAYS BE 0 FOR US!) */
+    std::uint8_t type = TYPE_RGB24;         /**< Type of image (ALWAYS 2 FOR RGB24!) */
+    std::uint8_t unused[5] = {};
+
+    std::uint16_t x_origin = 0;             /**< Co-ordinate for first X value */
+    std::uint16_t y_origin = 0;             /**< Co-ordinate for first Y value */
+    std::uint16_t width = 0;
+    std::uint16_t height = 0;
+    std::uint8_t bpp = 24;                  /**< Bits per pixel */
+    std::uint8_t flags = FLAGS_FLIP;
+};
+
+#pragma pack(pop)
+
+namespace
+{
+/**
+ *  Dump a texture to a TARGA/TGA file
+ *
+ *  @param width - The Width of this texture.
+ *  @param height - The Height of this texture.
+ *  @param data - Pixel data
+ *  @param bpp - Bitness of the texture to dump.
+ */
+void DumpRGB2Bitmap(std::uint32_t width, std::uint32_t height, std::vector<std::uint8_t>& data, std::uint8_t bpp = 24)
+{
+    tga_header header;
+    std::ofstream file("output.tga", std::ios::binary); // Open the stream for binary output
+
+    ASSERT(width <= std::numeric_limits<decltype(header.width)>::max());
+    ASSERT(height <= std::numeric_limits<decltype(header.height)>::max());
+
+    header.width        = static_cast<std::uint16_t>(width);
+    header.height       = static_cast<std::uint16_t>(height);
+
+    switch(bpp)
+    {
+    case 16:
+        header.bpp = 16;
+        break;
+    case 32:
+        header.bpp = 32;
+        break;
+    default:
+        break; // Do nothing as we have a default value of 32 (RGB).
```

Comment:
The header probably shouldn't set a default `bpp` either. This also removed the special-casing for `24`.